datatypes/src/raster/operations/scaling.rs

```diff
@@ -0,0 +1,631 @@
+    G: GridSize + Clone + Send + Sync,
+{
+    type Output = Grid<G, P>;
+
+    fn scale_elements_parallel(self, scale_with: P, offset_by: P, out_no_data: P) -> Self::Output {
+        self.map_elements_parallel(|p| p.scale(scale_with, offset_by), Some(out_no_data))
+    }
+}
+
+impl<P, G> ScaleElementsParallel<P> for GridOrEmpty<G, P>
+where
+    P: Scale + Copy + PartialEq + 'static + Send + Sync + DefaultNoDataValue,
+    G: GridSize + Clone + Send + Sync,
+    Grid<G, P>: ScaleElementsParallel<P, Output = Grid<G, P>>,
+{
+    type Output = GridOrEmpty<G, P>;
+
+    fn scale_elements_parallel(self, scale_with: P, offset_by: P, out_no_data: P) -> Self::Output {
+        match self {
+            GridOrEmpty::Grid(g) => {
+                GridOrEmpty::Grid(g.scale_elements_parallel(scale_with, offset_by, out_no_data))
+            }
+            GridOrEmpty::Empty(e) => GridOrEmpty::Empty(EmptyGrid::new(e.shape, out_no_data)),
+        }
+    }
+}
+
+impl<P> ScaleElementsParallel<P> for RasterTile2D<P>
+where
+    P: Scale + 'static + Copy + PartialEq + DefaultNoDataValue,
+    GridOrEmpty2D<P>: ScaleElementsParallel<P, Output = GridOrEmpty2D<P>>,
+{
+    type Output = RasterTile2D<P>;
+
+    fn scale_elements_parallel(self, scale_with: P, offset_by: P, out_no_data: P) -> Self::Output {
+        RasterTile2D {
+            grid_array: self
+                .grid_array
+                .scale_elements_parallel(scale_with, offset_by, out_no_data),
+            global_geo_transform: self.global_geo_transform,
+            properties: self.properties,
+            tile_position: self.tile_position,
+            time: self.time,
+        }
+    }
+}
+
+pub trait UnscaleElements<P> {
+    type Output;
+    fn unscale_elements(self, scale_with: P, offset_by: P, out_no_data: P) -> Self::Output;
+}
+
+impl<P, G> UnscaleElements<P> for Grid<G, P>
+where
+    P: Unscale + Copy + PartialEq + 'static + DefaultNoDataValue,
+    G: GridSize + Clone,
+{
+    type Output = Grid<G, P>;
+    fn unscale_elements(self, scale_with: P, offset_by: P, out_no_data: P) -> Self::Output {
+        self.map_elements(|p| p.unscale(scale_with, offset_by), Some(out_no_data))
+    }
+}
+impl<P, G> UnscaleElements<P> for GridOrEmpty<G, P>
+where
+    P: Unscale + Copy + PartialEq + 'static + DefaultNoDataValue,
+    G: GridSize + Clone,
+    Grid<G, P>: UnscaleElements<P, Output = Grid<G, P>>,
+{
+    type Output = GridOrEmpty<G, P>;
+    fn unscale_elements(self, scale_with: P, offset_by: P, out_no_data: P) -> Self::Output {
+        match self {
+            GridOrEmpty::Grid(g) => {
+                GridOrEmpty::Grid(g.unscale_elements(scale_with, offset_by, out_no_data))
+            }
+            GridOrEmpty::Empty(e) => GridOrEmpty::Empty(EmptyGrid::new(
+                e.shape,
+                e.no_data_value
+                    .unscale(scale_with, offset_by)
+                    .unwrap_or(P::DEFAULT_NO_DATA_VALUE),
```

Comment:
is this always meaningful or are there scenarios where we want to propagate this as an error?

Reply:
in this case i thing out of bounds values should cenvert to nodata

Reply:
[0f5694b](https://github.com/geo-engine/geoengine/pull/530/commits/0f5694b8b5b543dd192ecd08bf9651737e2f8afd)